File: leet-code/110.平衡二叉树.js
/*
 * @lc app=leetcode.cn id=110 lang=javascript
 *
 * [110] 平衡二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
  const getHeight = (node) => {
    if (!node) return 0

    // 后序遍历
    // 左
    const leftHight = getHeight(node.left)
    if (leftHight === -1) return -1

    // 右
    const rightHight = getHeight(node.right)
    if (rightHight === -1) return -1

    // 中
    if (Math.abs(leftHight - rightHight) > 1) return -1

    return Math.max(leftHight, rightHight) + 1
  }

  return getHeight(root) !== -1
};
// @lc code=end

File: leet-code-rs/src/solutions/110_平衡二叉树.rs
/*
 * @lc app=leetcode.cn id=110 lang=rust
 *
 * [110] 平衡二叉树
 */
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

#[allow(unused)]
impl Solution {
    pub fn max_depth(root: &Option<Rc<RefCell<TreeNode>>>) -> i32 {
        match root {
            Some(node) => {
                let left_depth = Self::max_depth(&node.borrow().left);
                if left_depth == -1 {
                    return -1;
                }
                let right_depth = Self::max_depth(&node.borrow().right);
                if right_depth == -1 {
                    return -1;
                }

                if (left_depth - right_depth).abs() > 1 {
                    return -1;
                }

                left_depth.max(right_depth) + 1
            }
            None => 0,
        }
    }
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let depth = Self::max_depth(&root);
        return depth != -1;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_is_balanced() {
        //assert_eq!(Solution::is_balanced(vec![]),[]);
        assert!(true)
    }
}
